jgrapht-core/src/main/java/org/jgrapht/alg/tour/NearestInsertionHeuristicTSP.java

```diff
@@ -0,0 +1,338 @@
+     *
+     * @param tourVertices Current tour vertices
+     * @param unvisited Set of unvisited vertices
+     * @param graph The graph
+     * @return Map storing closest unvisited vertex for each tour vertex
+     */
+    private Map<V, Closest<V, E>> getClosest(List<V> tourVertices, Set<V> unvisited, Graph<V, E> graph) {
+        return tourVertices.stream().collect(Collectors.toMap((v) -> v, (v) -> getClosest(v, unvisited, graph)));
+    }
+
+    /**
+     * Determines closest unvisited vertex to a vertex in the current tour
+     *
+     * @param tourVertex Vertex in the current tour
+     * @param unvisited Set of vertices not in the current tour
+     * @param graph The graph
+     * @return Closest unvisited vertex
+     */
+    private Closest<V, E> getClosest(V tourVertex, Set<V> unvisited, Graph<V, E> graph) {
+        V closest = null;
+        E closestEdge = null;
+        double minDist = Double.MAX_VALUE;
+        for (V unvisitedVertex : unvisited) {
+            double vDist = getDistance(graph, tourVertex, unvisitedVertex);
+            if (vDist < minDist) {
+                closest = unvisitedVertex;
+                closestEdge = graph.getEdge(tourVertex, unvisitedVertex);
+                minDist = vDist;
+            }
+        }
+        return new Closest<>(tourVertex, closest, closestEdge, minDist);
+    }
+
+    /**
+     * Update the Map storing closest unvisited vertex for each tour vertex
+     *
+     * @param currentClosest Map storing closest unvisited vertex for each tour
+     * vertex
+     * @param chosen Latest vertex added to tour
+     * @param unvisited Set of unvisited vertices
+     * @param graph The graph
+     * @return Updated map
+     */
+    private Map<V, Closest<V, E>> updateClosest(Map<V, Closest<V, E>> currentClosest, Closest<V, E> chosen, Set<V> unvisited, Graph<V, E> graph) {
+        // Update the set of unvisited vertices, and exit if none remain
+        unvisited.remove(chosen.getUnvisitedVertex());
+        if (unvisited.isEmpty()) {
+            return new HashMap<>(0);
+        }
+        // Update any entries impacted by the choice of new vertex
+        currentClosest.replaceAll((v, c) -> {
+            if (chosen.getTourVertex().equals(v) || chosen.getUnvisitedVertex().equals(c.getUnvisitedVertex())) {
+                return getClosest(v, unvisited, graph);
+            }
+            return c;
+        });
+        currentClosest.put(chosen.getUnvisitedVertex(), getClosest(chosen.getUnvisitedVertex(), unvisited, graph));
+        return currentClosest;
+    }
+
+    /**
+     * Chooses the closest unvisited vertex to the sub-tour
+     *
+     * @param closestVertices Map storing closest unvisited vertex for each tour
+     * vertex
+     * @return First result of sorting values
+     */
+    private Closest<V, E> chooseClosest(Map<V, Closest<V, E>> closestVertices) {
+        return closestVertices.values().stream().sorted().findFirst().get();
```

Comment:
It think `Collections.min(closestVertices.values());` is more efficient than sorting the entire `values`-Collection.